internal/gatewayapi/clienttrafficpolicy.go

```diff
@@ -329,7 +329,20 @@ func validatePortOverlapForClientTrafficPolicy(l *ListenerContext, xds *ir.Xds)
 	// IR must exist since we're past validation
 	if httpIR != nil {
 		if sameListeners := listenersWithSameHTTPPort(xds, httpIR); len(sameListeners) != 0 {
-			return fmt.Errorf("affects additional listeners: %s", strings.Join(sameListeners, ", "))
+			if attachedToGateway {
```

Comment:
this is now O(n^3)
can we instead hold a map containing listener name as the key and policy as the value ?
* if key is missing, add key, and return
* if key is present, policy is same, return
* if key is present, and a different policy, reject

this takes an approach of first policy wins, which imo is better than no policy winning

Reply:
> this takes an approach of first policy wins, which imo is better than no policy winning

"First policy wins" doesn't make sense. It's not a conflict between two policies where one can win and the other can lose, it's that translating a single policy (no other policies existing in the system) would result in it affecting multiple listeners including listeners that should not be affected by the policy.

Reply:
> this is now O(n^3)

Looking closely at the ` validatePortOverlapForClientTrafficPolicy` method, it seems to me that it's `O(n)` where `n` is the number of HTTP listeners attached to a Gateway XDS representation.


```golang

func validatePortOverlapForClientTrafficPolicy(l *ListenerContext, xds *ir.Xds, attachedToGateway bool) error {
	irListenerName := irHTTPListenerName(l)
	var httpIR *ir.HTTPListener

         // O(n) loop

	for _, http := range xds.HTTP { 
		if http.Name == irListenerName {
			httpIR = http
			break
		}
	}

	// IR must exist since we're past validation
	if httpIR != nil {
                //           listenersWithSameHTTPPort is O(n) and it is called once. 
                //  This is an 'if' statment, not a 'for' statement.
		if sameListeners := listenersWithSameHTTPPort(xds, httpIR); len(sameListeners) != 0 {


			if attachedToGateway {
				gatewayName := irListenerName[0:strings.LastIndex(irListenerName, "/")]
				conflictingListeners := []string{}

                                // sameListeners is an array with an upper bound of O(n)

				for _, currName := range sameListeners {
					if strings.Index(currName, gatewayName) != 0 {
						conflictingListeners = append(conflictingListeners, currName)
					}
				}
				if len(conflictingListeners) != 0 {
					return fmt.Errorf("affects additional listeners: %s", strings.Join(conflictingListeners, ", "))
				}
			} else {
				return fmt.Errorf("affects additional listeners: %s", strings.Join(sameListeners, ", "))
			}
		}
	}
	return nil
}

```

Reply:
@liorokman adding some reasons why we should consider first policy wins
* we already oldest config wins in most places https://github.com/envoyproxy/gateway/blob/6fa99be5a9a3e5d16294492162181ff64db67003/internal/gatewayapi/clienttrafficpolicy.go#L43
* this allows existing traffic to not get interrupted/dropped, ensures that is policy1 targets listener1 at t1 and policy2 targets listener2 at t2, listener1 is not affected

Reply:
@arkodg The bug occurs when there is exactly one policy. The scenario where there are two policies is out of scope for this issue. There's no way to implement "first policy wins", because "first policy wins" means "reject the policy provided".

Here's a recreation of the bug that this PR is trying to solve.:

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
    kind: Gateway
    metadata:
      name: gateway
    spec:
      gatewayClassName: envoy-gateway-class
      listeners:
        - name: http
          port: 80
          protocol: HTTP
          hostname: foo.example.com
          allowedRoutes:
            namespaces:
              from: Same
        - name: http2
          port: 80
          protocol: HTTP
          hostname: bar.example.com
          allowedRoutes:
            namespaces:
              from: Same
---
apiVersion: gateway.envoyproxy.io/v1alpha1
  kind: ClientTrafficPolicy
  metadata:
    name: target-gateway
  spec:
    targetRef:
      group: gateway.networking.k8s.io
      kind: Gateway
      name: gateway
      namespace: default
      sectionName: http
    path:
        escapedSlashesAction: RejectRequest
    timeout:
      http:
        requestReceivedTimeout: "5s"
      
```

The above single `ClientTrafficPolicy` tries to target only one listener - the one called `http` - and because both `http` and `http2` are non-TLS and on the same port, the `requestReceivedTimeout`  configured for `http2` will also be `5s`. Even worse, while `http` should have the `RejectRequest` action for escaped slashes, `http2` _should_ use the default value (`UnescapeAndForward`) but will instead be configured to `RejectRequest`.


Without this PR the translated IR will be correct - the various settings from the `ClientTrafficPolicy` will be set in the correct location - only on `http` and not on `http2`. The bug occurs because the correct IR is not translatable to valid XDS.